### ISSUE TITLE
More Performance Tweaks++

### ISSUE THREAD

**Issue by SteveTheTechie** [ehynds/jquery-ui-multiselect-widget#731]
I have previously developed some "no-jQuery", but "jQuery-like" DOM helper functions that allow doing efficient DOM operations in a sane way.  They are chainable functions which is why I say they are "jQuery-like".  I have adapted these for use in the widget and I have a version of _makeOption that is using these (it is faster).  They are below (at the bottom of this)...  Typical usage example:

```
var label = document.createElement('label')
	._setAttribute({ 'for': inputID,  'title': title})
	._addClass([isDisabled ? 'ui-state-disabled': null, isSelected && !isMultiple ? 'ui-state-active' : null, 'ui-corner-all'])
	._appendChild($input.get(0))
	._appendChild(span);

```
Do you have any objections to me using these to speed up the list creation?  Additionally, I want to get rid of the recursion in _buildOptions as it is preventing just doing one jQuery.append operation to add all list items from an array.  (one append from an array is faster)

```
	Element.prototype._setAttribute = function(attributes, value) {
		var elem = this;
		
		if (typeof attributes === "string") {
			elem.setAttribute(attributes, value);
			return elem;
		}		
		for (var name in attributes) {
			if (attributes[name] !== null)
				elem.setAttribute(name, attributes[name]);
		}
		return elem;
	};
	 
	Element.prototype._removeAttribute = function(attributes) {
		var elem = this;
		
		if (typeof attributes === "string") {
			elem.removeAttribute(attributes);
			return elem;
		}		
		var length = attributes.length;
		for(var x = 0; x < length; x++) {
			if (attributes[x])
				elem.removeAttribute(attributes[x]);
		}
		return elem;
	};
	
	Element.prototype._addClass = function(classes) {
		var elem = this;

		if (typeof classes === "string") 
			classes = classes.indexOf(' ') + 1 ? classes.split(' ') : [classes];		
		
		var length = classes.length;
		for(var x = 0; x < length; x++) {
			if (classes[x]) {
				if (elem.classList) 
					elem.classList.add(classes[x]);
				else 
					elem.className += ' ' + classes[x];
			}
		}
		return this;
	};
	
	Element.prototype._html = function(newhtml) {
		this.innerHTML = newhtml;
		return this;
	}
	
	Element.prototype._text = function(newtext) {
		this.textContent = newtext;
		return this;
	}
	
	Element.prototype._appendChild = function(childNode) {
		this.appendChild(childNode);
		return this;
	}
	
	Element.prototype._hasClass = function(className) {
		if (this.classList) return this.classList.contains(className);
		else return new RegExp('(^| )' + className + '( |$)', 'gi').test(this.className);
	};

```

**Comment by mlh758** [comment]
I created a gist for performance testing: https://gist.github.com/mlh758/b1b24931db7705441f9786bd68cf3569

Just drop it into the demos folder and open it in your browser. Version3 compared to Master so far gained about .3 (~2.5 to ~2.2) seconds across 5000 items with option groups in Chrome on my machine.

It would be nice to go faster, but I am not willing to modify the prototype on any core classes. That's asking for surprises when people pull the widget in. We can replace bits of the jQuery in stages to see what it gains us, and look for ways to reduce calls to append. We can also substitute append for just appending to the innerHTML and see what that gets us.

I'll have time to evaluate this stuff more carefully in the next few days 

*I hope

**Comment by mlh758** [comment]
I wouldn't be against removing the recursion, but it might be easier to just delay calling append instead once a larger chunk of HTML is built up.

I added it to support nested optgroups but that's not actually a thing so... now it's just easy to read.

**Comment by SteveTheTechie** [comment]
Ok, I will ditch the helper functions approach then and focus on doing what I can to just be more strategic w/ jQuery. 

I think the recursion can be ditched by just doing an inner each iteration for each optgroup.

I am already putting off doing the append, but I still end up doing multiple appends for the multiple optgroup case.  I think that the implication is that the only way to do one append w/ the recursion in place is if it is done in refresh() on the return from _buildOptions.  That is why I want to ditch the recursion.

**Comment by SteveTheTechie** [comment]
Reference the pull request.  The recursion was very simple to get rid of in _buildOptionList 

The only remaining possible performance tweak I can think of on the jQuery side is swapping out $().each() functions for native for loops.   However, I am on the fence, since I found conflicting information online about it.  It seems like it is worth doing for the $.each() generic iterator case, but maybe not for the $().each() jQuery object iterator case?

If you can think of any other jQuery related tweaks, please let me know.  Otherwise, I am going to move on to the width/height related stuff.

**Comment by SteveTheTechie** [comment]
Ok, So I did some timing using your gist... Thanks for creating that!  👍  I do see the 1-2 second time difference from the original widget code in the master branch... I can confirm that.

However, I don't really see any significant time difference between using $().each() vs for loops for this use case.  (maybe a tenth or two of a second, but it is debatable) I tend to think that plain for loops are better for plain objects, arrays, etc., but when iterating jQuery objects, using the jQuery each function seems to be a better way to go, since to use for loops, you still have to set up a number of variables and de-reference the jQuery object in the loop to get a specific element.  It ends up being a "wash", timewise, but using $().each() reads better.

I would like to formalize your gist into a demo / timer that handles multiple scenarios:  many options, many optgroups, and/or many instances.   It is the latter one that I am really interested in for my web app--that is why I am also tinkering w/ _create() and other functionality that affects how fast an instance can get created.

**Comment by mlh758** [comment]
I think you can fork the gist and play with it if you want to make more changes.

We can use the gist to get an actual measure of the changes we're making. Good to balance readability/simplicity with performance.

**Comment by SteveTheTechie** [comment]
There is one problem I found with your gist that has me a bit puzzled.  The multiselect will not open for me, at least in Chrome.  

I thought it was some bug I had introduced in the multiselect code, but all the unit tests pass, and all the demos work fine for me with the updated code.

What I discovered is that for some strange reason, when you are appending the elapsed time to the body, it is interfering with the events processing for some strange reason.
`document.getElementById('test').innerHTML += <h1>Widget time: ${(end - begin) / 1000} seconds</h1>;  (back-tics removed to allow proper gist formatting)`

If I remove this line or if I instead have it insert the HTML in an interior element w/ id="test", then all works fine... the multi-select opens fine, albeit slowly.

This is strange...

**Comment by mlh758** [comment]
I'll have to take a look when I get home today. I don't remember seeing any problems, but I also don't know if I actually tried to open the widget. I was just testing the load time.

I think the widget has some listeners on the body of the document, I'm curious if that's causing some kind of conflict. This issue also exists on master.

**Comment by SteveTheTechie** [comment]
Yes, that is what I was seeing also.  very strange

**Comment by mlh758** [comment]
Hahahah [woops](https://stackoverflow.com/questions/5113105/manipulating-innerhtml-removes-the-event-handler-of-a-child-element)

**Comment by SteveTheTechie** [comment]
Good lesson to keep in mind for the future.  ;-)  I think that the same thing can sometimes happen on divs when changing the innerHTML.   Sometimes I think it is better to use a target inner div or do an .appendChild()

**Comment by SteveTheTechie** [comment]
No jQuery helper functions that do not depend on Element.prototype:

```
_setAttribute: function(elem, attributes, value) {
	if (typeof attributes === "string") {
		elem.setAttribute(attributes, value);
		return elem;
	}		
	for (var name in attributes) {
		if (attributes[name] !== null)
			elem.setAttribute(name, attributes[name]);
	}
	return elem;
},
 
_removeAttribute: function(elem, attributes) {
	if (typeof attributes === "string") {
		elem.removeAttribute(attributes);
		return elem;
	}		
	var length = attributes.length;
	for(var x = 0; x < length; x++) {
		if (attributes[x])
			elem.removeAttribute(attributes[x]);
	}
	return elem;
},

_addClass: function(elem, classes) {
	if (typeof classes === "string") 
		classes = classes.indexOf(' ') + 1 ? classes.replace(/^\s+|\s+$/,'').split(/\s+/) : [classes];		
	
	var length = classes.length;
	for(var x = 0; x < length; x++) {
		if (classes[x]) {
			if (elem.classList) 
				elem.classList.add(classes[x]);
			else 
				elem.className += ' ' + classes[x];
		}
	}
	return elem;
},

_removeClass: function(elem, classes) {
	var elem = this;

	if (typeof classes === "string") 
		classes = classes.indexOf(' ') + 1 ? classes.replace(/^\s+|\s+$/,'').split(/\s+/) : [classes];		
	
	var length = classes.length;
	for(var x = 0; x < length; x++) {
		if (classes[x]) {
			if (elem.classList) 
				elem.classList.remove(classes[x]);
			else 
				elem.className = elem.className.replace(new RegExp('(^|\\b)' + classes[x].split(' ').join('|') + '(\\b|$)', 'gi'), ' ');
		}
	}
	return elem;
},

_hasClass: function(elem, className) {
	if (this.classList) 
		return this.classList.contains(className);
	else 
		return new RegExp('(^| )' + className + '( |$)', 'gi').test(this.className);
},

```

**Comment by SteveTheTechie** [comment]
As it turns out, I just reworked _makeOption w/ pure JS in my "pre-PR code": no jQuery, no helper functions.   It does shave off some more load time.

However, I tend to think that there are diminishing returns on trying to eek out more performance via converting jQuery to JS.  Also, jQuery does have its virtues, particularly for large sets of elements, so I am not sold on getting rid of much more of it.

I think there are probably some more re-flow lags, but those will probably get addressed better by reordering and restructuring the code a bit more rather than tinkering with specific lines of code and converting them to plain JS.

**Comment by mlh758** [comment]
Yeah, we're definitely looking at diminishing returns, and jQuery does provide some nice guarantees about cross browser behavior. It also keeps the code cleaner and easier to maintain. I'm also not a fan of re-inventing the wheel if we can avoid it.

If you still feel like there is a need to improve performance further it might be worth looking at the Chrome profiler to get an idea of where time is actually being spent during the load. Some of the earlier stuff you did was pretty straightforward, like switching out the element creation, but if more effort is going to go into this it would be better to start with some data on where it should be directed first.

**Comment by SteveTheTechie** [comment]
For your reference... it may be helpful to put marker in the Chrome Profiler's timeline:
[https://stackoverflow.com/questions/20941391/adding-marks-to-chromes-timeline-tool](https://stackoverflow.com/questions/20941391/adding-marks-to-chromes-timeline-tool)

**Comment by SteveTheTechie** [comment]
Basically, the essence of this has been implemented by #735, so I am closing this.
